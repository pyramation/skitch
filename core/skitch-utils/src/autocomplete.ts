import { filter } from 'fuzzy';

export const makeAutocompleteFunctionWithInput = (keys) => (answers: object, input: string) => {
  input = input || '';
  return new Promise(function(resolve) {
    setTimeout(function() {
      var fuzzyResult = filter(input, keys);
      resolve(
        fuzzyResult.map(function(el) {
          return el.original;
        })
      );
    }, 25);
  });
};
