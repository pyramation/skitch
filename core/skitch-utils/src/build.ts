import { listModules, getExtensionsAndModulesChanges } from './modules';
import { dirname, basename, resolve } from 'path';
import { skitchPath as sPath } from './paths';
import { readFileSync } from 'fs';

export const build = async argv => {
  const skitchPath = await sPath();
  const modules = await listModules();
  const modulesWithChanges = await getExtensionsAndModulesChanges('utils');

  const extensions = Object.keys(modules).map(key=>{
    const mod = modules[key];
    return {
      ...mod,
      sql: readFileSync(
        resolve(`${skitchPath}/${mod.path}/sql/${key}--${mod.version}.sql`)
      )
        .toString()
        .split('\n')
        .filter((l, i) => i !== 0)
        .join('\n')
    }
  });

  return extensions;
};
