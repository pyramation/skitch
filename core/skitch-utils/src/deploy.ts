import * as shell from 'shelljs';
import { resolve } from 'path';

import { listModules, getExtensionsAndModules } from './modules';

import { skitchPath } from './paths';

import { PGUSER, PGPASSWORD, PGHOST, PGPORT, PATH } from 'skitch-env';

const pg = require('pg');

// should we be parsing the plan file?
// currently assuming only extensions in control file...

export const deploy = async (name, database, opts) => {
  const path = await skitchPath();
  const modules = await listModules();
  const extensions = await getExtensionsAndModules(name);

  const pgPool = new pg.Pool({
    connectionString: `postgres://${PGUSER}:${PGPASSWORD}@${PGHOST}:${PGPORT}/${database}`
  });
  for (let n = 0; n < extensions.native.length; n++) {
    const extension = extensions.native[n];
    console.log(`CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`);
    await pgPool.query(
      `CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`
    );
  }
  pgPool.end();

  for (let n = 0; n < extensions.sqitch.length; n++) {
    const extension = extensions.sqitch[n];
    await deploy(extension, database, opts);
  }

  console.log(modules[name].path);
  console.log(`sqitch deploy db:pg:${database}`);
  shell.exec(`sqitch deploy db:pg:${database}`, {
    cwd: resolve(path, modules[name].path),
    env: {
      PGUSER,
      PGPASSWORD,
      PGHOST,
      PGPORT,
      PATH
    }
  });

  return extensions;
};
