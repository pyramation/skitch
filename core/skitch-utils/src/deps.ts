import { readFileSync, readFile } from 'fs';
import { basename, dirname, resolve, relative } from 'path';
import { sync as glob } from 'glob';
import { skitchPath } from './paths';
import { listModules } from './modules';

const makeKey = sqlmodule => '/deploy/' + sqlmodule + '.sql';

export const getDeps = async (packageDir) => {

  var external = [];

  var deps: { [type: string]: any } = {};
  var reg: { [type: string]: any } = {};

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(
    sqlmodule: string,
    resolved: string[],
    unresolved: string[]
  ) {
    unresolved.push(sqlmodule);
    let edges = deps[makeKey(sqlmodule)];
    if (!edges) {
      if (/:/.test(sqlmodule)) {
        external.push(sqlmodule);
        edges = deps[sqlmodule] = [];
      } else {
        throw new Error(`no module ${sqlmodule}`);
      }
    }

    for (var i = 0; i < edges.length; i++) {
      var dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(sqlmodule);
    var index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  var files = await glob(`${packageDir}/deploy/**/**.sql`);
  for (var i = 0; i < files.length; i++) {
    const data = readFileSync(files[i]);
    var lines = data.toString().split('\n');
    const key = '/' + relative(packageDir, files[i]);
    deps[key] = [];
    reg[key] = [];

    for (var j = 0; j < lines.length; j++) {
      var m = lines[j].match(/-- requires: (.*)/);
      if (m) {
        deps[key].push(m[1].trim());
      }

      // check only:
      var m2 = lines[j].match(/-- Deploy (.*) to pg/);
      if (m2) {
        if (key != makeKey(m2[1])) {
          throw new Error(
            'deployment script in wrong place or is named wrong internally' + m2
          );
        }
        reg[key].push(m2[1]);
      }
    }
  }

  let resolved: string[] = [];
  var unresolved: string[] = [];

  deps = Object.assign(
    {
      [makeKey('apps/index')]: Object.keys(deps)
        .filter(dep => dep.match(/^\/deploy\//))
        .map(dep => dep.replace(/^\/deploy\//, '').replace(/.sql$/, ''))
    },
    deps
  );

  dep_resolve('apps/index', resolved, unresolved);
  var index = resolved.indexOf('apps/index');
  resolved.splice(index);
  delete deps[makeKey('apps/index')];

  // move extensions up
  const extensions = resolved.filter(a => a.match(/^extensions/));
  const normalSql = resolved.filter(a => !a.match(/^extensions/));

  // resolved = useExtensions ? [...extensions, ...normalSql] : [...normalSql];
  resolved = [...extensions, ...normalSql];
  // resolved = [...normalSql];

  return {
    external,
    resolved,
    deps
  };

};
