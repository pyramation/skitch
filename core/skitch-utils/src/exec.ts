import { exec as shellExec } as shell from 'shelljs';
import { PGUSER, PGPASSWORD, PGHOST, PGPORT, PATH } from 'skitch-env';

export const execSync = (cmd, opts) => {
  shellExec(cmd, {
    env: {
      PGUSER,
      PGPASSWORD,
      PGHOST,
      PGPORT,
      PATH
    },
    ...opts
  });
};
