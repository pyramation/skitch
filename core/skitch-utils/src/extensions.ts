import { sync as glob } from 'glob';
import { sqitchPath as path } from './paths';
import { listModules } from './modules';
import { writeFileSync, readFileSync } from 'fs';
import { sluggify } from './utils';

export const getAvailableExtensions = async () => {
  let modules = await listModules();
  modules = Object.keys(modules).reduce(
    (m, v) => {
      if (m.includes(v)) return m;
      m.push(v);
      return m;
    },
    ['plpgsql', 'uuid-ossp', 'pgcrypto', 'plv8']
  );
  return modules;
};

export const getExtensionInfo = async () => {
  const sqitchPath = await path();
  const pkgPath = `${sqitchPath}/package.json`;
  const pkg = require(pkgPath);
  const extname = sluggify(pkg.name);
  const version = pkg.version;

  const Makefile = `${sqitchPath}/Makefile`;
  const controlFile = `${sqitchPath}/${extname}.control`;
  const sqlFile = `${extname}--${version}.sql`;
  return {
    extname,
    sqitchPath,
    version,
    Makefile,
    controlFile,
    sqlFile
  };
};

export const getInstalledExtensions = async () => {
  const info = await getExtensionInfo();

  let extensions;
  try {
    extensions = readFileSync(info.controlFile)
      .toString()
      .split('\n')
      .find(line => line.match(/^requires/))
      .split('=')[1]
      .split('\'')[1]
      .split(',')
      .map(a => a.trim());
  } catch (e) {
    throw new Error('missing requires from control files or bad syntax');
  }

  return extensions;
};

export const writeExtensionMakefile = async ({ path, extname, version }) => {
  writeFileSync(
    path,
    `EXTENSION = ${extname}
DATA = sql/${extname}--${version}.sql

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
  `
  );
};

export const writeExtensionControlFile = async ({
  path,
  extname,
  extensions,
  version
}) => {
  writeFileSync(
    path,
    `# ${extname} extension
comment = '${extname} extension'
default_version = '${version}'
module_pathname = '$libdir/${extname}'
requires = '${extensions.join(',')}'
relocatable = false
superuser = false
  `
  );
};

export const writeExtensions = async extensions => {
  const { controlFile: path, extname, version } = await getExtensionInfo();
  await writeExtensionControlFile({ path, extname, extensions, version });
};

export const writeExtensionsToEnv = async () => {
  const sqitchPath = await path();

  const controlFile = glob(`${sqitchPath}/*.control`);
  const envFile = glob(`${sqitchPath}/.env`);
  if (!controlFile || !controlFile.length) {
    throw new Error('no control file found!');
  }
  if (!envFile || !envFile.length) {
    throw new Error('no control file found!');
  }

  let extensions;
  try {
    extensions = readFileSync(controlFile[0])
      .toString()
      .split('\n')
      .find(line => line.match(/^requires/))
      .split('=')[1]
      .split('\'')[1]
      .split(',')
      .map(a => a.trim());
  } catch (e) {
    throw new Error('missing requires from control files or bad syntax');
  }

  let envs;
  try {
    envs = readFileSync(envFile[0])
      .toString()
      .split('\n')
      .reduce((m, line) => {
        line = (line || '').trim();
        if (/^#/.test(line)) return m;
        if (!line.length) return m;
        const parts = line.split('=');
        m[parts[0].trim()] = parts[1].trim();
        return m;
      }, {});
  } catch (e) {
    throw new Error('missing env files or bad syntax');
  }

  envs.PGEXTENSIONS = extensions.join(',');
  writeFileSync(
    envFile[0],
    Object.keys(envs).reduce((m, key) => {
      const value = envs[key];
      m = `${m}\n${key}=${value}`;
      return m;
    }, '')
  );
};
