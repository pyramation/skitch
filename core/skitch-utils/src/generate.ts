import { ChangePathArray } from 'skitch-types';

export const generate = async ({templates, template, templatePath, payload}) => {

  var params: Array<{ key: string; value: any }> = Object.keys(payload).reduce(
    (m: Array<{ key: string; value: any }>, v: string) => {
      if (payload[v] instanceof Array) {
        payload[v].forEach((value: string) => {
          m.push({
            key: v,
            value: value,
          });
        });
      } else {
        if (typeof payload[v] === 'boolean' && !payload[v]) {
          return m;
        }
        m.push({
          key: v,
          value: payload[v],
        });
      }
      return m;
    },
    []
  );
  var vars = params.map(obj => `--set ${obj.key}="${obj.value}"`).join(' ');

  let change = templates[template].change(payload);

  var reqd: ChangePathArray = [];

  let reqs: Array<ChangePathArray> = templates[template]
    .requires(payload)
    .filter((req: ChangePathArray) => {
      if (reqd.includes(req.join('/'))) {
        return false;
      }
      reqd.push(req.join('/'));
      return true;
    })
    .map((req: ChangePathArray) => {
      return `-r ${req.join('/')}`;
    })
    .join(' ');

  change = change.join('/');
  if (!change || change === '' || change === '/') {
    throw new Error('no change found!');
  }

  const cmd = [
    'sqitch',
    'add',
    change,
    '--template',
    template,
    '--template-directory',
    templatePath,
    '-n',
    `'add ${change}'`,
    vars,
    reqs,
  ].join(' ');

  return cmd;
};
