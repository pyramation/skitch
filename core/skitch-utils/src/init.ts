import 'skitch-template';
import { promisify } from 'util';
import { exec } from 'child_process';
import { sqitchPath as path, skitchPath as sPath } from './paths';
import { dirname, basename } from 'path';
import * as shell from 'shelljs';
import { writeFileSync } from 'fs';
const srcPath = dirname(require.resolve('skitch-template'));
import { makePlan } from './plans';

// import plan from './plan';

const makePackage = ({ name, description, author }) => {
  return {
    name,
    version: '0.0.1',
    description,
    author,
    private: true,
    scripts: {
      test: 'FAST_TEST=1 skitch-templatedb && jest',
      'test:watch': 'FAST_TEST=1 jest --watch',
    },
    devDependencies: {
      '@types/jest': '21.1.0',
      '@types/node': '8.0.0',
      'babel-cli': '6.24.1',
      'babel-jest': '20.0.3',
      'babel-preset-react-app': '3.0.0',
      dotenv: '5.0.1',
      jest: '20.0.4',
    },
    dependencies: {
      pg: '6.4.0',
      'pg-promise': '6.10.3',
      'skitch-testing': 'latest',
      uuid: '3.1.0',
    },
  };
};

const sluggify = (text) => {
  return text.toString().toLowerCase().trim()
    .replace(/\s+/g, '-')           // Replace spaces with -
    .replace(/&/g, '-and-')         // Replace & with 'and'
    .replace(/[^\w\-]+/g, '')       // Remove all non-word chars
    .replace(/\-\-+/g, '-');        // Replace multiple - with single -
}

export const init = async ({ name, description, author, extensions }) => {

  await sPath();

  // init sqitch

  const cmd = ['sqitch', 'init', name, '--engine', 'pg'].join(' ');
  await promisify(exec)(cmd.trim());

  // now we have a sqitch path!

  const sqitchPath = await path();
  const pkg = makePackage({ name, description, author });

  // initialize template
  shell.cp('-r', `${srcPath}/sqitch/*`, `${sqitchPath}/`);
  shell.cp('-r', `${srcPath}/sqitch/.*`, `${sqitchPath}/`);

  shell.mkdir('-p', `${sqitchPath}/sql`);
  const extname = sluggify(name);

  writeFileSync(`${sqitchPath}/Makefile`, `EXTENSION = ${extname}
DATA = sql/${extname}--0.0.1.sql

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
  `);

  writeFileSync(`${sqitchPath}/${extname}.control`, `# ${extname} extension
comment = '${description}'
default_version = '0.0.1'
module_pathname = '$libdir/${extname}'
requires = '${extensions.join(',')}'
relocatable = false
superuser = false
  `);

  writeFileSync(`${sqitchPath}/package.json`, JSON.stringify(pkg, null, 2));

  const settings = {
    name,
    // projects: true
  };

  // const plan = await makePlan(sqitchPath, settings);
  // fs.writeFileSync(`${sqitchPath}/sqitch.plan`, plan);

  console.log(`

        |||
       (o o)
   ooO--(_)--Ooo-


✨ ${name} created!

Now try this:

skitch generate

`);
};

export const initSkitch = async () => {

  const dir = process.cwd();
  shell.cp('-r', `${srcPath}/skitch/*`, `${dir}/`);
  shell.cp('-r', `${srcPath}/skitch/.*`, `${dir}/`);

  console.log(`

        |||
       (o o)
   ooO--(_)--Ooo-


✨ Great work! Now, try this:

cd packages/
mkdir myfirstmodule
cd myfirstmodule/
skitch init
`);

};
