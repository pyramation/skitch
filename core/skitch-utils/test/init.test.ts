import { init } from '../src/init';
import { sync as mkdirp } from 'mkdirp';
import { sync as glob } from 'glob';
import { sync as rimraf } from 'rimraf';
import { writeFileSync } from 'fs';

const TMPDIR = process.env.TMPDIR;
const rnd = () =>
  Math.random()
    .toString(36)
    .substring(2, 15) +
  Math.random()
    .toString(36)
    .substring(2, 15);

describe('deps', () => {
  let dir, projDir;
  beforeEach(async () => {
    dir = TMPDIR + '/' + rnd();
    projDir = dir + '/myproject';
    mkdirp(projDir);
    process.chdir(dir);
    await initSkitch();
    process.chdir(projDir);
  });
  afterEach(() => {
    rimraf(dir);
  });
  describe('skitch path', () => {
    it('skitch init', async () => {
      process.chdir(dir);

      const files = glob('**');
      expect(files).toEqual(["myproject", "skitch.json"]);
      const hidden = glob('.*');
      expect(hidden).toEqual(['.travis.yml']);
    });
    it('sqitch init', async () => {
      await init({
        name: 'myproject',
        description: 'my amazing project',
        author: 'dan@example.com',
        extensions: ['plpgsql', 'citext']
      });

      const files = glob('**');
      expect(files).toEqual([
        'deploy',
        'docker-compose.yml',
        'Makefile',
        'myproject.control',
        'package.json',
        'revert',
        'sqitch.conf',
        'sqitch.plan',
        'sql',
        'test',
        'test/utils',
        'test/utils/index.js',
        'verify'
      ]);
      const hidden = glob('.*');
      expect(hidden).toEqual(['.babelrc', '.env', '.travis.yml']);
    });
  });
});
