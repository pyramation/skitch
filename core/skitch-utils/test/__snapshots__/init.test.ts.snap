// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deps skitch path skitch init 1`] = `
Array [
  "bootstrap-roles.sql",
  "docker-compose.yml",
  "Makefile",
  "myproject",
  "package.json",
  "packages",
  "packages/install.sh",
  "readme.md",
  "skitch.json",
]
`;

exports[`deps skitch path skitch init 2`] = `
Array [
  ".eslintrc.json",
  ".travis.yml",
]
`;

exports[`deps skitch path sqitch init 1`] = `
Array [
  "deploy",
  "Makefile",
  "myproject.control",
  "package.json",
  "revert",
  "sqitch.conf",
  "sqitch.plan",
  "sql",
  "test",
  "test/utils",
  "test/utils/index.js",
  "verify",
]
`;

exports[`deps skitch path sqitch init 2`] = `
Array [
  ".babelrc",
  ".env",
]
`;
